### ISSUE TITLE
Confusing error message when uncancelling subscription

### ISSUE THREAD

**Issue by beccagaspard** [killbill/killbill#512]
When I try to uncancel a subscription that is already in a cancelled state I get this error message: "Error: Subscription {id} was not in a cancelled state: Failed to uncancel plan". 

https://github.com/killbill/killbill/blob/master/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L348

I'm not sure what the correct behavior is here..  Should users be able to uncancel subscriptions that are already cancelled, or can they only uncancel future cancellations?  If the latter, maybe the error message should be clearer.  :)

**Comment by sbrossie** [comment]
@beccagaspard The behavior is the later:  they can only uncancel future (pending) cancellations.

Agree on fixing the message. Technically this is not really an api change so could go on 0.16.3.

**Comment by beccagaspard** [comment]
I was looking into this a little more, and it seems that currently you are only allowed to uncancel future cancellations where the entitlement is not yet cancelled.  If you cancel a subscription with an effective date in the future and choose entitlementPolicy = IMMEDIATE, you will not be able to uncancel.  Is this how it should work, or should users still be able to uncancel when the entitlement is already cancelled?

**Comment by sbrossie** [comment]
@beccagaspard You are raising an interesting (product) semantics question. Let me provide some background. The future cancellation scenarios can be summarized in the following 3 use cases:

1. Same (future) `effectiveDate` for both entitlement and subscription (billing piece). This is for when billing needs to reflect entitlement (service).
2. Entitlement `effectiveDate`  > subscription `effectiveDate`. I actually don't know of a good use case for that one, although we support it.
3. Entitlement `effectiveDate`  < subscription `effectiveDate`. This is a common use case to stop service right away but avoid any pro-ration credit.

The scenario you describe is a subset of 3 where  Entitlement `effectiveDate` = NOW. In that case, the cancellation has been reached (user does not have access to service anymore), but to avoid pro-ration credit, subscription `effectiveDate`  is still in the future.

Also cancellation is a terminal state, so once it has been reached the only way to restart service/billing would be to create a new subscription. However you are correct that in that example creating a new subscription would end up with double billing for the period [NOW, subscription `effectiveDate`]. Is that the issue you are having?

**Comment by beccagaspard** [comment]
Thanks for the clarification - I think I was actually (unintentionally) running into 2 - I cancelled for a future date and didn't set the useRequestedDateForBilling flag to true, resulting in a billing end date of today and an entitlement end date of today + 7.  The state of the subscription showed as ACTIVE, so I figured I could still uncancel it, but got the same error message as in my 1st message.  You're right, can't really think of a good use case for this scenario, just ran into it accidentally.  

So just to make sure I understand, you can only uncancel subscriptions when both the billing end date and entitlement end date are in the future?  Lots of possible scenarios... thanks for your help! :)

**Comment by sbrossie** [comment]
From [reading our code](https://github.com/killbill/killbill/blob/work-for-release-0.16.3/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L358), the current implementation does the following:
* If subscription (billing) is already cancelled =>  throw Exception
* If not allow the uncancel to go through -- although we also [throw if we discover along the way that entitlement is not already cancelled or future cancelled](https://github.com/killbill/killbill/blob/work-for-release-0.16.3/entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlement.java#L379). 

So to summarize with what i said previously i believe we support all 3 cases except 2/ when subscription cancellation `effectiveDate` <= NOW.

**Comment by sbrossie** [comment]
@beccagaspard Should i close the issue, or is there anything we should be doing?

**Comment by sbrossie** [comment]
@beccagaspard Actually re-reading the issue i see this an error code message issue.

Will move the issue to 0.17.0 (this is not really an api change but still oblige us to rebuild api, where the [ErrorCode](https://github.com/killbill/killbill-api/blob/master/src/main/java/org/killbill/billing/ErrorCode.java) live.
